python/cudf/udf_cpp/shim.cu

```diff
@@ -720,3 +722,49 @@ make_definition_corr(BlockCorr, int64, int64_t);
 
 #undef make_definition_corr
 }
+
+/*
+NRT CUDA functions
+*/
+
+// Only used to allocate the right amount of space, see below
+struct meminfo_and_str {
+  NRT_MemInfo mi;
+  udf_string st;
+};
+
+/*
+Create a new MemInfo object holding the reference count of a udf_string. When returning
+new strings, shim functions expect a pointer to a stack allocated buffer into which it
+will construct the udf_string it returns. Since one can not safely build a MemInfo object
+around this stack memory, we store a copy of the udf_string itself next to the MemInfo
+which manages the lifetime of the udf_string.
+*/
+extern "C" __device__ int meminfo_from_new_udf_str(void** nb_retval, void* udf_str)
```

Comment:
I see the main purpose of this function as "Move a stack-allocated Managed UDF String to a new heap allocation, without moving the underlying string data". At this point I'm still a little fuzzy as to why the stack-allocated version reaching the end of its life doesn't delete the underlying string data.

Reply:
Is that because `udf_str_dtor` only gets called for the heap-allocated Managed UDF Strings and not the stack-allocated ones?

Reply:
`udf_str_dtor` should only ever be called by a MemInfo when the refcount of a string reaches zero. I'll write something a little more detailed around the lifecycle of a string on this PR.